feathr_project/feathr/_preprocessing_pyudf_manager.py

```diff
@@ -0,0 +1,135 @@
+            if preprocessing_func:
+                need_preprocessing = True
+
+                self.persist_pyspark_udf_to_file(preprocessing_func, local_workspace_dir)
+                feature_names = [feature.name for feature in anchor.features]
+                features_with_preprocessing = features_with_preprocessing + feature_names
+                feature_names.sort()
+                string_feature_list = ','.join(feature_names)
+                feature_names_to_func_mapping[string_feature_list] = anchor.source.preprocessing
+
+        if not need_preprocessing:
+            return
+
+        self.write_feature_names_to_udf_name_file(feature_names_to_func_mapping, local_workspace_dir)
+
+        # Save necessary preprocessing-related metadata locally in your workspace
+        # Typically it's used as a metadata for join/gen job to figure out if there is preprocessing UDF
+        # exist for the features requested
+        feathr_pyspark_metadata_abs_path = os.path.join(local_workspace_dir, FEATHR_PYSPARK_METADATA)
+        file = open(feathr_pyspark_metadata_abs_path, 'wb')
+        pickle.dump(features_with_preprocessing, file)
+        file.close()
+
+    def persist_pyspark_udf_to_file(self, user_func, local_workspace_dir):
+        # Some basic imports will be provided
+        udf_source_code = inspect.getsourcelines(user_func)[0]
+        lines = []
+        lines = lines + PROVIDED_IMPORTS
+        lines = lines + udf_source_code
+        lines.append('\n')
+
+        client_udf_repo_path = os.path.join(local_workspace_dir, FEATHR_CLIENT_UDF_FILE_NAME)
+
+        # the directory may actually not exist yet, so create the directory first
+        file_name_start = client_udf_repo_path.rfind("/")
+        if file_name_start > 0:
+            dir_name = client_udf_repo_path[:file_name_start]
+            Path(dir_name).mkdir(parents=True, exist_ok=True)
+
+        if Path(client_udf_repo_path).is_file():
+            with open(client_udf_repo_path, "a") as handle:
+                print("".join(lines), file=handle)
+        else:
+            with open(client_udf_repo_path, "w") as handle:
+                print("".join(lines), file=handle)
+
+    def write_feature_names_to_udf_name_file(self, feature_names_to_func_mapping, local_workspace_dir):
+        """Persist feature names(sorted) of an anchor to the corresponding preprocessing function name to source path
+        under the local_workspace_dir.
+        """
+        # indent in since python needs correct indentation
+        # Don't change the indentation
+        tm = Template("""
+feature_names_funcs = {
+{% for key, value in func_maps.items() %}
+    "{{key}}" : {{value.__name__}},
+{% endfor %}
+}
+        """)
+        new_file = tm.render(func_maps=feature_names_to_func_mapping)
+
+        full_file_name = os.path.join(local_workspace_dir, FEATHR_CLIENT_UDF_FILE_NAME)
+        with open(full_file_name, "a") as text_file:
```

Comment:
Would this append mode result in super long files (with previous unused functions or even conflict names etc.) if we have the same workspace and same file name? maybe we should clean it up a bit before each run?